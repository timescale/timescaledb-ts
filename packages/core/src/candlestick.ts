import {
  CandlestickAggregateOptions,
  CandlestickAggregateOptionsSchema,
  TimeRange,
  WhereClause,
} from '@timescaledb/schemas';
import { buildWhereClause, escapeIdentifier } from '@timescaledb/utils';

export class CandlestickAggregateBuilder {
  private statements: string[] = [];
  private options: CandlestickAggregateOptions;
  private tableName: string;

  constructor(tableName: string, options: CandlestickAggregateOptions) {
    this.tableName = tableName;
    this.options = CandlestickAggregateOptionsSchema.parse(options);
  }

  private buildWhere(where: WhereClause, paramOffset = 1): { sql: string; params: any[] } {
    if (!where) {
      return { sql: '', params: [] };
    }

    const { sql, params } = buildWhereClause(where, paramOffset);

    return { sql: ` AND ${sql}`, params };
  }

  public build({ where, range }: { where?: WhereClause; range?: TimeRange } = {}): { sql: string; params: any[] } {
    const tableName = escapeIdentifier(this.tableName);
    const timeColumn = escapeIdentifier(this.options.time_column!);
    const priceColumn = escapeIdentifier(this.options.price_column);
    const volumeColumn = this.options.volume_column ? escapeIdentifier(this.options.volume_column) : null;
    const interval = '$1::interval';

    this.statements = [];

    this.statements.push(`SELECT`);
    this.statements.push(`  time_bucket(${interval}, ${timeColumn}) as bucket_time,`);
    this.statements.push(
      `  open(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as open,`,
    );
    this.statements.push(
      `  high(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as high,`,
    );
    this.statements.push(
      `  low(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as low,`,
    );
    this.statements.push(
      `  close(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as close,`,
    );
    this.statements.push(
      `  open_time(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as open_time,`,
    );
    this.statements.push(
      `  high_time(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as high_time,`,
    );
    this.statements.push(
      `  low_time(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as low_time,`,
    );
    this.statements.push(
      `  close_time(candlestick_agg(${timeColumn}, ${priceColumn}${volumeColumn ? `, ${volumeColumn}` : ''})) as close_time`,
    );

    if (volumeColumn) {
      this.statements.push(`  ,volume(candlestick_agg(${timeColumn}, ${priceColumn}, ${volumeColumn})) as volume,`);
      this.statements.push(`  vwap(candlestick_agg(${timeColumn}, ${priceColumn}, ${volumeColumn})) as vwap`);
    }

    this.statements.push(`FROM ${tableName}`);

    const params: (string | Date)[] = [this.options.bucket_interval || '1 hour'];
    if (range) {
      params.push(range.start, range.end);
    }

    const whereClause = this.buildWhere(where as WhereClause, params.length + 1);
    const timeRangeWhere = range ? `WHERE ${timeColumn} >= $2 AND ${timeColumn} <= $3` : '';
    const fullWhere = timeRangeWhere + whereClause.sql;

    if (fullWhere) {
      this.statements.push(fullWhere);
    }

    params.push(...whereClause.params);

    this.statements.push(`GROUP BY bucket_time`);
    this.statements.push(`ORDER BY bucket_time ASC;`);

    return {
      sql: this.statements.join('\n'),
      params,
    };
  }
}

export interface CandlestickMetadata {
  propertyName: string;
  timeColumn?: string;
  priceColumn?: string;
  volumeColumn?: string;
  sourceColumn?: string;
}

export class CandlestickBuilder {
  static generateSQL(metadata: CandlestickMetadata, isRollup: boolean = false): string {
    if (!metadata) return '';

    const targetColumn = escapeIdentifier(metadata.propertyName);

    if (isRollup) {
      if (!metadata.sourceColumn) {
        throw new Error('source_column must be specified for candlestick rollups');
      }
      const sourceColumn = escapeIdentifier(metadata.sourceColumn);
      return `rollup(${sourceColumn}) as ${targetColumn}`;
    }

    if (!metadata.timeColumn || !metadata.priceColumn) {
      throw new Error('time_column and price_column must be specified for candlestick aggregation');
    }

    const args = [escapeIdentifier(metadata.timeColumn), escapeIdentifier(metadata.priceColumn)];

    if (metadata.volumeColumn) {
      args.push(escapeIdentifier(metadata.volumeColumn));
    }

    return `candlestick_agg(${args.join(', ')}) as ${targetColumn}`;
  }
}
