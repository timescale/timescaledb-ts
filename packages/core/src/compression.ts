import { CompressionSelect, CompressionSelectSchema, CreateHypertableOptions } from '@timescaledb/schemas';
import { escapeLiteral } from '@timescaledb/utils';

class CompressionStatsBuilder {
  private statements: string[] = [];

  private name: string;
  private options: CompressionSelect;

  constructor(name: string, select: CompressionSelect) {
    this.name = name;
    this.options = select;
  }

  public build(): string {
    this.statements.push(`SELECT`);

    const columns: string[] = [];

    if (this.options.total_chunks) {
      columns.push(`COALESCE(total_chunks, 0)::integer as total_chunks`);
    }

    if (this.options.compressed_chunks) {
      columns.push(`COALESCE(number_compressed_chunks, 0)::integer as compressed_chunks`);
    }

    if (columns.length === 0) {
      this.statements.push(`*`);
    } else {
      this.statements.push(columns.join(',\n'));
    }

    const literalName = escapeLiteral(this.name);

    this.statements.push(`FROM hypertable_compression_stats(${literalName});`);

    return this.statements.join('\n');
  }
}

export class CompressionBuilder {
  private name: string;
  private options: CreateHypertableOptions;

  constructor(name: string, options: CreateHypertableOptions) {
    this.name = name;
    this.options = options;
  }

  public stats(statOpts: { select: CompressionSelect }): CompressionStatsBuilder {
    const select = CompressionSelectSchema.parse(statOpts.select);
    return new CompressionStatsBuilder(this.name, select);
  }
}
