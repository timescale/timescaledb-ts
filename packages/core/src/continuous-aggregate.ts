import { CreateContinuousAggregateOptions, CreateContinuousAggregateOptionsSchema } from '@timescaledb/schemas';
import { escapeIdentifier, escapeLiteral } from '@timescaledb/utils';

class ContinuousAggregateInspectBuilder {
  constructor(private name: string) {}

  public build(): string {
    const literalName = escapeLiteral(this.name);

    return `SELECT EXISTS (
      SELECT FROM timescaledb_information.continuous_aggregates
      WHERE view_name = ${literalName}
    ) as hypertable_exists;`;
  }
}

class ContinuousAggregateUpBuilder {
  constructor(
    private name: string,
    private source: string,
    private options: CreateContinuousAggregateOptions,
  ) {}

  private generateAggregate(config: { type: string; column?: string; column_alias: string }): string {
    const alias = escapeIdentifier(config.column_alias);

    switch (config.type) {
      case 'count':
        return `COUNT(*) as ${alias}`;
      case 'count_distinct': {
        if (!config.column) {
          throw new Error('Column is required for count_distinct aggregate');
        }
        const column = escapeIdentifier(config.column);
        return `COUNT(DISTINCT ${column}) as ${alias}`;
      }
      case 'sum': {
        if (!config.column) {
          throw new Error('Column is required for sum aggregate');
        }
        const sumColumn = escapeIdentifier(config.column);
        return `SUM(${sumColumn}) as ${alias}`;
      }
      case 'avg': {
        if (!config.column) {
          throw new Error('Column is required for avg aggregate');
        }
        const avgColumn = escapeIdentifier(config.column);
        return `AVG(${avgColumn}) as ${alias}`;
      }
      case 'min': {
        if (!config.column) {
          throw new Error('Column is required for min aggregate');
        }
        const minColumn = escapeIdentifier(config.column);
        return `MIN(${minColumn}) as ${alias}`;
      }
      case 'max': {
        if (!config.column) {
          throw new Error('Column is required for max aggregate');
        }
        const maxColumn = escapeIdentifier(config.column);
        return `MAX(${maxColumn}) as ${alias}`;
      }
      default:
        throw new Error(`Unsupported aggregate type: ${config.type}`);
    }
  }

  private generateSelect(): string {
    const timeColumn = escapeIdentifier(this.options.time_column);
    const interval = escapeLiteral(this.options.bucket_interval);
    const sourceName = escapeIdentifier(this.source);

    const aggregates = Object.entries(this.options.aggregates).map(([, config]) => {
      return this.generateAggregate(config);
    });

    return `
      SELECT
        time_bucket(${interval}, ${timeColumn}) as bucket,
        ${aggregates.join(',\n        ')}
      FROM ${sourceName}
      GROUP BY bucket
    `;
  }

  public getRefreshPolicy(): string | null {
    if (!this.options.refresh_policy) return null;

    const policy = this.options.refresh_policy;
    const viewName = escapeLiteral(this.name);
    return `SELECT add_continuous_aggregate_policy(${viewName},
      start_offset => INTERVAL ${escapeLiteral(policy.start_offset)},
      end_offset => INTERVAL ${escapeLiteral(policy.end_offset)},
      schedule_interval => INTERVAL ${escapeLiteral(policy.schedule_interval)}
    );`;
  }

  public build(): string {
    const viewName = escapeIdentifier(this.name);
    return `CREATE MATERIALIZED VIEW ${viewName} WITH (timescaledb.continuous) AS ${this.generateSelect()} WITH NO DATA;`;
  }
}

class ContinuousAggregateDownBuilder {
  constructor(
    private name: string,
    private options: CreateContinuousAggregateOptions,
  ) {}

  public build(): string[] {
    const statements: string[] = [];
    const viewName = this.name;

    if (this.options.refresh_policy) {
      statements.push(`SELECT remove_continuous_aggregate_policy(${escapeLiteral(viewName)}, if_exists => true);`);
    }

    statements.push(`DROP MATERIALIZED VIEW IF EXISTS ${escapeIdentifier(viewName)};`);

    return statements;
  }
}

export class ContinuousAggregate {
  private name: string;
  private source: string;
  private options: CreateContinuousAggregateOptions;

  constructor(name: string, source: string, options: CreateContinuousAggregateOptions) {
    this.name = name;
    this.source = source;
    this.options = CreateContinuousAggregateOptionsSchema.parse(options);
  }

  public up(): ContinuousAggregateUpBuilder {
    return new ContinuousAggregateUpBuilder(this.name, this.source, this.options);
  }

  public down(): ContinuousAggregateDownBuilder {
    return new ContinuousAggregateDownBuilder(this.name, this.options);
  }

  public inspect(): ContinuousAggregateInspectBuilder {
    return new ContinuousAggregateInspectBuilder(this.name);
  }
}
