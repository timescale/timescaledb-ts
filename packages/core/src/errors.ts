export enum HypertableErrors {
  NAME_REQUIRED = 'Hypertable name is required',
  OPTIONS_REQUIRED = 'Hypertable options are required',
  INVALID_OPTIONS = 'Invalid hypertable options',
}
