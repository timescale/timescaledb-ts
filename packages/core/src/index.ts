export const name = '@timescaledb/core';
