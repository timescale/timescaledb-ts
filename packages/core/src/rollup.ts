import { RollupConfig, RollupRule } from '@timescaledb/schemas';
import { escapeIdentifier, escapeLiteral } from '@timescaledb/utils';
import { CandlestickBuilder, CandlestickMetadata } from './candlestick';

export interface RollupMetadata {
  candlestick?: CandlestickMetadata;
  rollupRules: Array<RollupRule>;
}

class RollupInspectBuilder {
  constructor(private config: RollupConfig) {}

  public build(): string {
    if (!this.config || !this.config.rollupOptions) {
      throw new Error('Invalid rollup configuration');
    }

    const sourceView = escapeLiteral(this.config.rollupOptions.sourceView);
    const rollupView = escapeLiteral(this.config.rollupOptions.name);

    return `
      SELECT 
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = ${sourceView}
        ) as source_view_exists,
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = ${rollupView}
        ) as rollup_view_exists;
    `;
  }
}

class RollupUpBuilder {
  private statements: string[] = [];

  constructor(private config: RollupConfig) {}

  private buildRollupSelect(metadata: { candlestick?: CandlestickMetadata; rollupRules: Array<RollupRule> }): string {
    const selectStatements: string[] = [];

    const bucketInterval = escapeLiteral(this.config.rollupOptions.bucketInterval);
    const bucketColumn = escapeIdentifier('bucket');
    selectStatements.push(`time_bucket(${bucketInterval}::interval, ${bucketColumn}) AS ${bucketColumn}`);

    if (metadata.candlestick) {
      const rollupSql = CandlestickBuilder.generateSQL(metadata.candlestick, true);
      if (rollupSql) {
        selectStatements.push(rollupSql);
      }
    }

    const rollupSelects = metadata.rollupRules.map((rule) => {
      const sourceColumn = escapeIdentifier(rule.sourceColumn);
      const targetColumn = escapeIdentifier(rule.targetColumn || 'bucket');

      if (rule.aggregateType) {
        switch (rule.aggregateType) {
          case 'sum':
            return `sum(${sourceColumn}) as ${targetColumn}`;
          case 'avg':
            return `avg(${sourceColumn}) as ${targetColumn}`;
          default:
            return `rollup(${sourceColumn}) as ${targetColumn}`;
        }
      }

      return `${rule.rollupFn || 'rollup'}(${sourceColumn}) as ${targetColumn}`;
    });

    selectStatements.push(...rollupSelects);

    const sourceView = escapeIdentifier(this.config.rollupOptions.sourceView);
    return `
      SELECT
        ${selectStatements.join(',\n        ')}
      FROM ${sourceView}
      GROUP BY 1${this.config.rollupOptions.materializedOnly ? ' WITH ' : ' WITH NO '}DATA;
    `;
  }

  public build(metadata: RollupMetadata): string {
    const viewName = escapeIdentifier(this.config.rollupOptions.name);
    this.statements.push(
      `CREATE MATERIALIZED VIEW ${viewName}
      WITH (timescaledb.continuous) AS ${this.buildRollupSelect(metadata)}`,
    );

    return this.statements.join('\n');
  }

  public getRefreshPolicy(): string | null {
    if (!this.config.continuousAggregateOptions.refresh_policy) return null;

    const policy = this.config.continuousAggregateOptions.refresh_policy;
    const viewName = escapeLiteral(this.config.rollupOptions.name);

    return `SELECT add_continuous_aggregate_policy(${viewName},
      start_offset => INTERVAL ${escapeLiteral(policy.start_offset)},
      end_offset => INTERVAL ${escapeLiteral(policy.end_offset)},
      schedule_interval => INTERVAL ${escapeLiteral(policy.schedule_interval)}
    );`;
  }
}

class RollupDownBuilder {
  constructor(private config: RollupConfig) {}

  public build(): string[] {
    const statements: string[] = [];
    const viewName = this.config.rollupOptions.name;

    if (this.config.continuousAggregateOptions.refresh_policy) {
      statements.push(`SELECT remove_continuous_aggregate_policy(${escapeLiteral(viewName)}, if_exists => true);`);
    }

    statements.push(`DROP MATERIALIZED VIEW IF EXISTS ${escapeIdentifier(viewName)};`);

    return statements;
  }
}

export class RollupBuilder {
  constructor(private config: RollupConfig) {
    if (!config || !config.rollupOptions) {
      throw new Error('Invalid rollup configuration');
    }
  }

  public up(): RollupUpBuilder {
    return new RollupUpBuilder(this.config);
  }

  public down(): RollupDownBuilder {
    return new RollupDownBuilder(this.config);
  }

  public inspect(): RollupInspectBuilder {
    return new RollupInspectBuilder(this.config);
  }
}
