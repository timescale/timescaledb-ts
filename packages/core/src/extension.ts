import { CreateExtensionOptions, CreateExtensionOptionsSchema } from '@timescaledb/schemas';
import { ExtensionErrors } from './errors';

class ExtensionUpBuilder {
  private options?: CreateExtensionOptions;
  private statements: string[] = [];

  constructor(options?: CreateExtensionOptions) {
    this.options = options;
  }

  public build(): string {
    const stmt = `CREATE EXTENSION IF NOT EXISTS timescaledb${this?.options?.should_cascade ? ' CASCADE' : ''};`;
    this.statements.push(stmt);

    return this.statements.join('\n');
  }
}

class ExtensionDownBuilder {
  private options?: CreateExtensionOptions;
  private statements: string[] = [];

  constructor(options?: CreateExtensionOptions) {
    this.options = options;
  }

  public build(): string {
    const stmt = `DROP EXTENSION IF EXISTS timescaledb${this?.options?.should_cascade ? ' CASCADE' : ''};`;
    this.statements.push(stmt);

    return this.statements.join('\n');
  }
}

export class Extension {
  private options?: CreateExtensionOptions;

  constructor(options?: CreateExtensionOptions) {
    if (options) {
      try {
        this.options = CreateExtensionOptionsSchema.parse(options);
      } catch (error) {
        const e = error as Error;
        throw new Error(ExtensionErrors.INVALID_OPTIONS + ' ' + e.message);
      }
    }
  }

  public up(): ExtensionUpBuilder {
    return new ExtensionUpBuilder(this.options);
  }

  public down(): ExtensionDownBuilder {
    return new ExtensionDownBuilder(this.options);
  }
}
