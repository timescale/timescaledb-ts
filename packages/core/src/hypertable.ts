import { CreateHypertableOptions, CreateHypertableOptionsSchema } from '@timescaledb/schemas';
import { HypertableErrors } from './errors';
import { escapeIdentifier, escapeLiteral, validateIdentifier } from '@timescaledb/utils';

class HypertableUpBuilder {
  private options: CreateHypertableOptions;
  private name: string;
  private statements: string[] = [];

  constructor(name: string, options: CreateHypertableOptions) {
    this.name = name;
    this.options = options;
  }

  public build(): string {
    const tableName = escapeIdentifier(this.name);

    this.statements.push(
      `SELECT create_hypertable(${escapeLiteral(this.name)}, by_range(${escapeLiteral(this.options.by_range.column_name)}));`,
    );

    if (this.options.compression?.compress) {
      const orderBy = escapeIdentifier(this.options.compression.compress_orderby);
      const segmentBy = escapeIdentifier(this.options.compression.compress_segmentby);

      const alter = `ALTER TABLE ${tableName} SET (
        timescaledb.compress,
        timescaledb.compress_orderby = ${orderBy},
        timescaledb.compress_segmentby = ${segmentBy}
      );`;
      this.statements.push(alter);

      if (this.options.compression.policy) {
        const interval = escapeLiteral(this.options.compression.policy.schedule_interval);
        const policy = `SELECT add_compression_policy(${escapeLiteral(this.name)}, INTERVAL ${interval});`;
        this.statements.push(policy);
      }
    }

    return this.statements.join('\n');
  }
}

class HypertableDownBuilder {
  private name: string;
  private options: CreateHypertableOptions;
  private statements: string[] = [];

  constructor(name: string, options: CreateHypertableOptions) {
    this.name = name;
    this.options = options;
  }

  public build(): string {
    const tableName = escapeIdentifier(this.name);
    const literalName = escapeLiteral(this.name);

    if (this.options.compression?.compress) {
      this.statements.push(`ALTER TABLE ${tableName} SET (timescaledb.compress = false);`);
    }

    if (this.options.compression?.policy) {
      this.statements.push(`SELECT remove_compression_policy(${literalName}, if_exists => true);`);
    }

    this.statements.push(`SELECT drop_chunks(${literalName}, NOW()::timestamp without time zone);`);

    return this.statements.join('\n');
  }
}

export class Hypertable {
  private options: CreateHypertableOptions;
  private name: string;

  constructor(name: string, options: CreateHypertableOptions) {
    if (!name) {
      throw new Error(HypertableErrors.NAME_REQUIRED);
    }

    try {
      validateIdentifier(name, true);
      this.name = name;
    } catch (error) {
      throw new Error(HypertableErrors.INVALID_NAME + ' ' + (error as Error).message);
    }

    if (!options) {
      throw new Error(HypertableErrors.OPTIONS_REQUIRED);
    }

    try {
      this.options = CreateHypertableOptionsSchema.parse(options);
    } catch (error) {
      const e = error as Error;
      throw new Error(HypertableErrors.INVALID_OPTIONS + ' ' + e.message);
    }
  }

  public up(): HypertableUpBuilder {
    return new HypertableUpBuilder(this.name, this.options);
  }

  public down(): HypertableDownBuilder {
    return new HypertableDownBuilder(this.name, this.options);
  }
}
