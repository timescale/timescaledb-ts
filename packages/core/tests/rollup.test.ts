import { describe, it, expect } from '@jest/globals';
import { RollupBuilder } from '../src/rollup';
import { RollupConfig, RollupFunctionType } from '@timescaledb/schemas';

describe('RollupBuilder', () => {
  const baseConfig: RollupConfig = {
    continuousAggregateOptions: {
      name: 'daily_rollup',
      bucket_interval: '1 day',
      time_column: 'time',
      refresh_policy: {
        start_offset: '30 days',
        end_offset: '1 day',
        schedule_interval: '1 day',
      },
    },
    rollupOptions: {
      name: 'daily_rollup',
      sourceView: 'hourly_metrics',
      bucketInterval: '1 day',
      rollupRules: [
        {
          rollupFn: RollupFunctionType.Rollup,
          sourceColumn: 'percentile_hourly',
          targetColumn: 'percentile_daily',
        },
      ],
      materializedOnly: false,
    },
  };

  describe('up()', () => {
    it('should generate basic rollup SQL', () => {
      const builder = new RollupBuilder(baseConfig);
      const sql = builder.up().build();
      expect(sql).toMatchSnapshot();
    });

    it('should generate SQL with materialized-only option', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          materializedOnly: true,
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.up().build();
      expect(sql).toMatchSnapshot();
    });

    it('should generate SQL with multiple rollup rules', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          rollupRules: [
            {
              rollupFn: RollupFunctionType.Rollup,
              sourceColumn: 'percentile_hourly',
              targetColumn: 'percentile_daily',
            },
            {
              rollupFn: RollupFunctionType.Mean,
              sourceColumn: 'response_time',
              targetColumn: 'mean_response_time',
            },
          ],
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.up().build();
      expect(sql).toMatchSnapshot();
    });

    it('should handle column names with special characters', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          rollupRules: [
            {
              rollupFn: RollupFunctionType.Rollup,
              sourceColumn: 'percentile"hourly',
              targetColumn: 'percentile"daily',
            },
          ],
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.up().build();
      expect(sql).toMatchSnapshot();
    });

    describe('refresh policy', () => {
      it('should generate refresh policy SQL', () => {
        const builder = new RollupBuilder(baseConfig);
        const sql = builder.up().getRefreshPolicy();
        expect(sql).toMatchSnapshot();
      });

      it('should handle null refresh policy', () => {
        const config = {
          ...baseConfig,
          continuousAggregateOptions: {
            ...baseConfig.continuousAggregateOptions,
            refresh_policy: undefined,
          },
        };
        const builder = new RollupBuilder(config);
        const sql = builder.up().getRefreshPolicy();
        expect(sql).toBeNull();
      });

      it('should properly escape interval values', () => {
        const config = {
          ...baseConfig,
          continuousAggregateOptions: {
            ...baseConfig.continuousAggregateOptions,
            refresh_policy: {
              start_offset: "30 days'--injection",
              end_offset: '1 day',
              schedule_interval: '1 day',
            },
          },
        };
        const builder = new RollupBuilder(config);
        const sql = builder.up().getRefreshPolicy();
        expect(sql).toMatchSnapshot();
      });
    });
  });

  describe('down()', () => {
    it('should generate drop statements', () => {
      const builder = new RollupBuilder(baseConfig);
      const sql = builder.down().build();
      expect(sql).toMatchSnapshot();
    });

    it('should generate drop statements without refresh policy', () => {
      const config = {
        ...baseConfig,
        continuousAggregateOptions: {
          ...baseConfig.continuousAggregateOptions,
          refresh_policy: undefined,
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.down().build();
      expect(sql).toMatchSnapshot();
    });

    it('should properly escape view names with special characters', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          name: 'daily"rollup',
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.down().build();
      expect(sql).toMatchSnapshot();
    });
  });

  describe('inspect()', () => {
    it('should generate basic inspect SQL', () => {
      const builder = new RollupBuilder(baseConfig);
      const sql = builder.inspect().build();
      expect(sql).toMatchSnapshot();
    });

    it('should properly escape view names with special characters', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          name: 'daily"rollup',
          sourceView: 'hourly"metrics',
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.inspect().build();
      expect(sql).toMatchSnapshot();
    });

    it('should handle schema qualified names', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          name: 'public.daily_rollup',
          sourceView: 'public.hourly_metrics',
        },
      };
      const builder = new RollupBuilder(config);
      const sql = builder.inspect().build();
      expect(sql).toMatchSnapshot();
    });
  });

  describe('error handling', () => {
    it('should validate required source columns', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          rollupRules: [
            {
              rollupFn: RollupFunctionType.Rollup,
              // @ts-ignore
              sourceColumn: undefined,
              targetColumn: 'rolled_up_value',
            },
          ],
        },
      };
      // @ts-ignore
      const builder = new RollupBuilder(config as RollupConfig);
      expect(() => builder.up().build()).toThrow();
    });

    it('should validate bucket interval', () => {
      const config = {
        ...baseConfig,
        rollupOptions: {
          ...baseConfig.rollupOptions,
          // @ts-ignore
          bucketInterval: '',
        },
      };
      const builder = new RollupBuilder(config as RollupConfig);
      expect(() => builder.up().build()).toThrow();
    });
  });
});
