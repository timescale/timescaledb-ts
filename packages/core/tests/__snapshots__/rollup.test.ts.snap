// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RollupBuilder SQL Generation down() should escape special characters in view names 1`] = `
[
  "SELECT remove_continuous_aggregate_policy('daily"rollup.view', if_exists => true);",
  "DROP MATERIALIZED VIEW IF EXISTS "daily""rollup.view";",
]
`;

exports[`RollupBuilder SQL Generation down() should generate drop statements with refresh policy 1`] = `
[
  "SELECT remove_continuous_aggregate_policy('daily_rollup', if_exists => true);",
  "DROP MATERIALIZED VIEW IF EXISTS "daily_rollup";",
]
`;

exports[`RollupBuilder SQL Generation down() should generate drop statements without refresh policy 1`] = `
[
  "SELECT remove_continuous_aggregate_policy('daily_rollup', if_exists => true);",
  "DROP MATERIALIZED VIEW IF EXISTS "daily_rollup";",
]
`;

exports[`RollupBuilder SQL Generation inspect() should escape special characters in view names 1`] = `
"
      SELECT 
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = 'source"view'
        ) as source_view_exists,
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = 'my"rollup.view'
        ) as rollup_view_exists;
    "
`;

exports[`RollupBuilder SQL Generation inspect() should generate inspect SQL 1`] = `
"
      SELECT 
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = 'hourly_metrics'
        ) as source_view_exists,
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = 'daily_rollup'
        ) as rollup_view_exists;
    "
`;

exports[`RollupBuilder SQL Generation inspect() should handle schema qualified names 1`] = `
"
      SELECT 
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = 'public.hourly_metrics'
        ) as source_view_exists,
        EXISTS (
          SELECT FROM information_schema.views 
          WHERE table_schema = 'public' 
          AND table_name = 'public.daily_rollup'
        ) as rollup_view_exists;
    "
`;

exports[`RollupBuilder SQL Generation refresh policy should escape interval values in refresh policy 1`] = `
"SELECT add_continuous_aggregate_policy('daily_rollup',
      start_offset => INTERVAL '30 days''--injection',
      end_offset => INTERVAL '1 day',
      schedule_interval => INTERVAL '1 day'
    );"
`;

exports[`RollupBuilder SQL Generation refresh policy should generate refresh policy SQL 1`] = `
"SELECT add_continuous_aggregate_policy('daily_rollup',
      start_offset => INTERVAL '30 days',
      end_offset => INTERVAL '1 day',
      schedule_interval => INTERVAL '1 day'
    );"
`;

exports[`RollupBuilder SQL Generation up() should generate SQL with materialized-only option 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "bucket") AS "bucket",
      rollup("percentile_hourly") as "percentile_daily",
      sum("total_hourly") as "total_daily"
      FROM "hourly_metrics"
      GROUP BY 1 WITH DATA;
    "
`;

exports[`RollupBuilder SQL Generation up() should generate basic rollup SQL 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "bucket") AS "bucket",
      rollup("percentile_hourly") as "percentile_daily",
      sum("total_hourly") as "total_daily"
      FROM "hourly_metrics"
      GROUP BY 1 WITH NO DATA;
    "
`;

exports[`RollupBuilder SQL Generation up() should handle different aggregate types 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "bucket") AS "bucket",
      sum("value") as "sum_value",
      avg("value") as "avg_value"
      FROM "hourly_metrics"
      GROUP BY 1 WITH NO DATA;
    "
`;

exports[`RollupBuilder SQL Generation up() should handle different bucket columns 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "custom_bucket") AS "custom_bucket",
      rollup("percentile_hourly") as "percentile_daily",
      sum("total_hourly") as "total_daily"
      FROM "hourly_metrics"
      GROUP BY 1 WITH NO DATA;
    "
`;

exports[`RollupBuilder SQL Generation up() should handle group columns 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "bucket") AS "bucket",
      "symbol" as "symbol",
      "exchange" as "exchange",
      rollup("percentile_hourly") as "percentile_daily",
      sum("total_hourly") as "total_daily"
      FROM "hourly_metrics"
      GROUP BY 1, "symbol", "exchange" WITH NO DATA;
    "
`;

exports[`RollupBuilder SQL Generation up() should handle special characters in column names 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "bucket") AS "bucket",
      rollup("percentile""hourly") as "percentile""daily"
      FROM "hourly_metrics"
      GROUP BY 1 WITH NO DATA;
    "
`;

exports[`RollupBuilder SQL Generation up() should include candlestick columns 1`] = `
"CREATE MATERIALIZED VIEW "daily_rollup"
      WITH (timescaledb.continuous) AS 
      SELECT
        time_bucket('1 day'::interval, "bucket") AS "bucket",
      rollup("candlestick") as "candlestick",
      rollup("percentile_hourly") as "percentile_daily",
      sum("total_hourly") as "total_daily"
      FROM "hourly_metrics"
      GROUP BY 1 WITH NO DATA;
    "
`;
